firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/internal/SharedPreferencesUtil.java

```diff
@@ -0,0 +1,213 @@
+
+    if (modelHash == null || modelHash.isEmpty()) {
+      // no model hash means no download in progress
+      return null;
+    }
+
+    long fileSize =
+        getSharedPreferences()
+            .getLong(String.format(DOWNLOADING_MODEL_SIZE_PATTERN, persistenceKey, modelName), 0);
+
+    long id =
+        getSharedPreferences()
+            .getLong(String.format(DOWNLOADING_MODEL_ID_PATTERN, persistenceKey, modelName), 0);
+
+    return new CustomModel(modelName, id, fileSize, modelHash);
+  }
+
+  /**
+   * The information about the new custom model download that need to be stored.
+   *
+   * @param customModel custom model details to be stored.
+   */
+  public synchronized void setDownloadingCustomModelDetails(@NonNull CustomModel customModel) {
+    String modelName = customModel.getName();
+    String modelHash = customModel.getModelHash();
+    long downloadId = customModel.getDownloadId();
+    long modelSize = customModel.getSize();
+    getSharedPreferences()
+        .edit()
+        .putString(
+            String.format(DOWNLOADING_MODEL_HASH_PATTERN, persistenceKey, modelName), modelHash)
+        .putLong(
+            String.format(DOWNLOADING_MODEL_SIZE_PATTERN, persistenceKey, modelName), modelSize)
+        .putLong(String.format(DOWNLOADING_MODEL_ID_PATTERN, persistenceKey, modelName), downloadId)
+        // The following assumes the download will finish before the system reboots.
+        // If not, the download duration won't be correct, which isn't critical.
+        .putLong(
+            String.format(DOWNLOAD_BEGIN_TIME_MS_PATTERN, persistenceKey, modelName),
+            SystemClock.elapsedRealtime())
+        .apply();
+  }
+
+  /**
+   * The information about a completed custom model download. Updates the local model information
+   * and clears the download details associated with this model.
+   *
+   * @param customModel custom model details to be stored.
+   */
+  public synchronized void setUploadedCustomModelDetails(@NonNull CustomModel customModel)
+      throws IllegalArgumentException {
+    Long id = customModel.getDownloadId();
+    // only call when download is completed and download id is reset to 0;
+    if (!id.equals(0L)) {
+      throw new IllegalArgumentException("Only call when Custom model has completed download.");
+    }
+
+    String modelName = customModel.getName();
+    String hash = customModel.getModelHash();
+    long size = customModel.getSize();
+    String filePath = customModel.getLocalFilePath();
+    getSharedPreferences()
+        .edit()
+        .putString(String.format(LOCAL_MODEL_HASH_PATTERN, persistenceKey, modelName), hash)
+        .putLong(String.format(LOCAL_MODEL_FILE_SIZE_PATTERN, persistenceKey, modelName), size)
+        .putString(
+            String.format(LOCAL_MODEL_FILE_PATH_PATTERN, persistenceKey, modelName), filePath)
+        .apply();
+
+    clearDownloadingModelDetails(customModel.getName());
```

Comment:
Would it be useful to do the clearing as part of the change above? If yes, then you can pass the editor to `clearDownloadingModelDetails` and use it to add the remove statements in a single `apply()` call.

Reply:
great idea.